libraries/AP_InertialSensor/AP_InertialSensor.cpp

```diff
@@ -1301,67 +1293,68 @@ void AP_InertialSensor::update(void)
+            } 
+            else if (AP_HAL::millis() - _startup_ms > 2000) {
                 _startup_error_counts_set = true;
             }
         }
 
         for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
-            if (_accel_error_count[i] < _accel_startup_error_count[i]) {
+            if (_accel_error_count[i] <= _accel_startup_error_count[i]) {
                 _accel_startup_error_count[i] = _accel_error_count[i];
             }
-            if (_gyro_error_count[i] < _gyro_startup_error_count[i]) {
-                _gyro_startup_error_count[i] = _gyro_error_count[i];
-            }
-        }
-
-        // adjust health status if a sensor has a non-zero error count
-        // but another sensor doesn't.
-        bool have_zero_accel_error_count = false;
-        bool have_zero_gyro_error_count = false;
-        for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
-            if (_accel_healthy[i] && _accel_error_count[i] <= _accel_startup_error_count[i]) {
-                have_zero_accel_error_count = true;
+            else {
+                _accel_healthy[i] = false;
             }
-            if (_gyro_healthy[i] && _gyro_error_count[i] <= _gyro_startup_error_count[i]) {
-                have_zero_gyro_error_count = true;
+            
+            if (_gyro_error_count[i] <= _gyro_startup_error_count[i]) {
+                _gyro_startup_error_count[i] = _gyro_error_count[i];
             }
-        }
-
-        for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
-            if (_gyro_healthy[i] && _gyro_error_count[i] > _gyro_startup_error_count[i] && have_zero_gyro_error_count) {
-                // we prefer not to use a gyro that has had errors
+            else {
                 _gyro_healthy[i] = false;
             }
-            if (_accel_healthy[i] && _accel_error_count[i] > _accel_startup_error_count[i] && have_zero_accel_error_count) {
-                // we prefer not to use a accel that has had errors
-                _accel_healthy[i] = false;
-            }
         }
 
         // set primary to first healthy accel and gyro
-        for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
-            if (_gyro_healthy[i] && _use[i]) {
-                _primary_gyro = i;
-                break;
-            }
-        }
-        for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
-            if (_accel_healthy[i] && _use[i]) {
-                _primary_accel = i;
-                break;
-            }
-        }
+        _primary_gyro = find_primary_gyro();
+        _primary_accel = find_primary_accel();
     }
 
     // apply notch filter to primary gyro
-    _gyro[_primary_gyro] = _notch_filter.apply(_gyro[_primary_gyro]);
-    
+    _gyro[_primary_gyro] = _notch_filter.apply(_gyro[_primary_gyro]);    
     _last_update_usec = AP_HAL::micros();
     
     _have_sample = false;
 }
 
+
+int8_t AP_InertialSensor::find_primary_gyro() const {
+    if(_primary_gyro >= 0 && _gyro_healthy[_primary_gyro]) {
```

Comment:
should be faster

Reply:
But it's a functional change. Not only it doesn't check the `_use` parameter, but it's staying with the same primary, even when there is an healthy one with a smaller index - not intending to say that's an incorrect change, just pointing it out.

Reply:
I made _primary_gyro an int initialized with -1;
If it is bigger it means the loop found assigned already. 
The _use array is changed nowhere, so we don't need to constantly check.